### PR TITLE
enh: Added support for RealBinOp in abs

### PR DESCRIPTION
**Pull request by jinangshah21**
It is just enhancement of #5528 . It now supports RealBinOp also.

### CODE REVIEW

**Comment by jinangshah21** [comment]
> Considering that the test case added here works with `--experimental-simplifier` but doesn't work without it in the current main, I'm also interested in now extracting the current failing third party MRE with the CI failure.

It works without `--experimental-simplifier` in current main.

**Comment by gxyd** [comment]
> It works without `--experimental-simplifier` in current main.

I tried the updated test that you've added in this PR, and ran like this with current main:
```console
> main_lfortran integration_tests/array_op_08.f90                                                 lf 11:26:34 AM
Internal Compiler Error: Unhandled exception
Traceback (most recent call last):
LCompilersException: Expression type, f32[3] not yet supported
```

That's the error I get. While with `--experimental-simplifier` it succeeds:
```console
> main_lfortran integration_tests/array_op_08.f90 --experimental-simplifier                       lf 11:28:35 AM
T    T    T
T    T    T
```

**Comment by jinangshah21** [comment]
> > It works without `--experimental-simplifier` in current main.
> 
> I tried the updated test that you've added in this PR, and ran like this with current main:
> 
> ```
> > main_lfortran integration_tests/array_op_08.f90                                                 lf 11:26:34 AM
> Internal Compiler Error: Unhandled exception
> Traceback (most recent call last):
> LCompilersException: Expression type, f32[3] not yet supported
> ```
> 
> That's the error I get. While with `--experimental-simplifier` it succeeds:
> 
> ```
> > main_lfortran integration_tests/array_op_08.f90 --experimental-simplifier                       lf 11:28:35 AM
> T    T    T
> T    T    T
> ```

In My Local main it is working : 
```console
(lf) jinu2110@JNSLAP:/mnt/c/lfortran$ lfortran a.f90
T    T    T    T
```

**Comment by gxyd** [comment]
What's your version of LFortran? Mine is:
```console
> lfortran --version
LFortran version: 0.42.0-249-gaed88c2ec-dirty
Platform: macOS ARM
LLVM: 11.1.0
Default target: arm64-apple-darwin23.6.0
```

**Comment by gxyd** [comment]
We anyways should extract the MRE from fastGPT for the current failing third party test, to ensure that it's a part of our integration tests.

**Comment by gxyd** [comment]
> I am not able to identify which example is failing in FastGPT.

I don't think it will be very easy to extract the MRE, you'll have to use the technique of commenting out to reduce the size of code in fastGPT and that will help us find the MRE. (As this is a runtime failure, it may not be straight forward)

**Comment by gxyd** [comment]
> In My Local main it is working :
> 
> ```
> (lf) jinu2110@JNSLAP:/mnt/c/lfortran$ lfortran integration_tests/array_op_08.f90
> T    T    T
> T    T    T
> ```

Also, if the currently added test in this PR already passes with the current *main*, why do we need the change in this PR?

**Comment by jinangshah21** [comment]
> > In My Local main it is working :
> > ```
> > (lf) jinu2110@JNSLAP:/mnt/c/lfortran$ lfortran integration_tests/array_op_08.f90
> > T    T    T
> > T    T    T
> > ```
> 
> Also, if the currently added test in this PR already passes with the current _main_, why do we need the change in this PR?

Oh Sorry, with changes of this PR it will pass.

**Comment by gxyd** [comment]
> Oh Sorry, with changes of this PR it will pass.

Ah, I see.

I guess then the first thing now we need to do is to extract the MRE from fastGPT, it can be a little challenging to extract but definitely be worth the effort.

**Comment by jinangshah21** [comment]
ASR after `array_op` pass for MRE of intrinsics_57 : 
1) Without your change : 
```fortran
(Assignment
      (ArrayItem
          (Var 2 res)
          [(()
          (Var 2 __1_t2)
          ())]
          (Integer 4)
          RowMajor
          ()
      )
      (IntrinsicElementalFunction
          Merge
          [(IntegerConstant 1 (Integer 4) Decimal)
          (ArrayItem
              (Var 2 a)
              [(() 
              (Var 2 __1_v1)
              ())]
              (Integer 4)
              RowMajor
              ()
          )
          (ArrayItem
              (Var 2 __libasr__created__var__0__string_comp_op_res)
              [(()
              (Var 2 __1_v1)
              ())]
              (Logical 4)
              RowMajor
              ()
          )]
          0
          (Integer 4)
          ()
      )
      ()
  )
```
2) With above suggested change :
```fortran 
(Assignment
      (ArrayItem
          (Var 2 __libasr__created__var__0__elemental_func_call_res)
          [(()
          (Var 2 __1_t2)
          ())]
          (Integer 4)
          RowMajor
          ()
      )
      (IntrinsicElementalFunction
          Merge
          [(IntegerConstant 1 (Integer 4) Decimal)
          (ArrayItem
              (Var 2 a)
              [(()
              (Var 2 __1_v1)
              ())]
              (Integer 4)
              RowMajor
              ()
          )
          (ArrayItem
              (Var 2 res)
              [(()
              (Var 2 __1_v1)
              ())]
              (Integer 4)
              RowMajor
              ()
          )]
          0
          (Integer 4)
          ()
      )
      ()
  )
```

**Comment by Pranavchiku** [comment]
Try to see if you can fix the CI, then we can review this PR thoroughly

**Comment by jinangshah21** [comment]
```fortran
program intrinsics_57
implicit none

character(len=:), allocatable :: shorts(:)
character(len=:), allocatable :: long
integer :: a(10), i, res(10)
do i = 1, 10
    a(i) = i
end do

allocate(character(len=6) :: shorts(10))
allocate(character(len=5) :: long)
shorts(1:10:2) = "shorts"
shorts(2:10:2) = "long"
long = "long"

res = merge(1, a, shorts == long)
end program

```

**Review by certik — APPROVED** [review]
I think that looks good.

File: src/libasr/pass/array_op.cpp [hunk 1]
```diff
@@ -1568,8 +1568,12 @@ class ReplaceArrayOp: public ASR::BaseExprReplacer<ReplaceArrayOp> {
...
         bool at_least_one_array = false;
         for( size_t iarg = 0; iarg < x->n_args; iarg++ ) {
-            array_mask[iarg] = ASRUtils::is_array(
-                ASRUtils::expr_type(x->m_args[iarg]));
+            ASR::expr_t* arg = x->m_args[iarg];
+            if (ASR::is_a<ASR::ArrayPhysicalCast_t>(*arg)) {
+                arg = ASR::down_cast<ASR::ArrayPhysicalCast_t>(arg)->m_arg;
+            }
+            array_mask[iarg] = ASRUtils::is_array(ASRUtils::expr_type(arg)) || ASR::is_a<ASR::IntrinsicArrayFunction_t>(*arg) ||
+                        ASR::is_a<ASR::RealBinOp_t>(*arg) || ASR::is_a<ASR::IntegerBinOp_t>(*arg);
```
Comment (Pranavchiku) [inline src/libasr/pass/array_op.cpp:1576]: Brilliant! This looks good.

**Pull request merged**